### DISCUSSION TITLE
Better support for Map, Set and iterables

### DISCUSSION THREAD

**Discussion by etodanik** [remeda/remeda#972] [General]
Most of the functions that target `Array` would work great if they could natively support `Set` and anything that conforms to the iterable standard. 

Almost anything that works `Object` should also work great with both `Set` and `Map`. What's more - by working with an array of tuples ([[key, value], [key, value]]) - you could create pipes that modify keys and values simultaneously, so in theory you could map() a Map into a new Map :) 

Working with any iterable would also allow to pipe generator functions and various readable streams that conform to the iterator standard. 

Is `remeda` interested in such functionality? It feels like a natural modern evolution for a utility belt library, and so far `remeda` is the best one I could find for TS.

I'd love to participate in creating PR for some of this stuff but would first wanna know if it's of interest.

**Comment by TkDodo** [comment]
I think it sounds interesting! Does that mean we could also map from e.g. an `Array` into an `Object`? That's a quite common use-case where we now have an extra function `mapToObj` and `flatMapToObj` that takes an Array, expects a tuple to be returned and creates an object out of it.

Can you show how an example api for this would look like, and would this be a breaking change?

Thanks for helping out and being interested in remeda 💖

**Comment by CarlosFdez** [comment]
A bit of a resurrection, but one example we ran into is that we sometimes have a object that maps keys -> localization strings, and sometimes we want a subset of that key->localization mapping. But if the keys list is a Set, we can't use it in R.pick() directly, and have to do `R.pick(someObject, [...someSet])`.

It could be a case where when it makes sense, an argument could take an Iterable instead of an array, and I don't *think* that would be too much of a lift.

**Comment by eranhirsch** [comment]
I had this in mind too, but the thing is that for both `map` and `filter` which are most likely the most common use-cases for iterating over containers, JS doesn't offer native implementations, meaning you are effectively breaking down the container into an array of entries (or values, or keys) and then performing any transformation on that array.

This means that having a function that would for example map on a Map would effectively:
1. create an array of entries
2. perform the mapping on that array
3. create a *new* Map from that array.
Let's say you want to map and then filter, you are now creating a redundant temporary map, and going through the whole process of creating the entries array for it. This can't be good, considering the alternative would be to stay in "array-land" until the final step that you build the map again.

The conclusion from this is that beyond making `toPairs`, `values` and `keys` work for Maps and Sets; and maybe providing a function `mapFromEntries` and `setFromValues` (but those could just be arrow functions `(entries) => new Map(entries)` inside the pipe).

Of course, there are some quality of life functions we can provide, I have a utility I use in my projects which I call `isRefinedPair` which I use with filters on arrays of entries to allow me to write stuff like `filter(isRefinedPair({ value: isDefined }))`. I'd be happy to send a PR to include it.

I'm also happy to talk about specific use-cases where addressing Sets and Maps directly is useful.

**Comment by somnicattus** [comment]
Hello.

I published remeda-compatible npm package rotery, which expands remeda with Iterators and AsyncIterators support.

https://github.com/somnicattus/rotery

**Comment by olsonpm** [comment]
Anyone looking for a utility library which approaches data structures generically - my solution is Common FP

https://common-fp.org

```
$ npm install common-fp

# if using typescript, additionally run
$ npm install common-fp-types
```

It covers the builtin structures, mostly object, array, Map, Set, and a few utilities for other types like string, number and Date

I think the simplest example is [mapValues](https://common-fp.org/docs/mapValues) which maps over the values of an array, object, Map or Set and returns a new instance of whatever type you passed in.

There's also [a playground to try it out](https://common-fp.org/try-it#editorData=eyJjb2RlIjoiaW1wb3J0IHsgbWFwVmFsdWVzIH0gZnJvbSAnY29tbW9uLWZwJ1xuXG5jb25zdCBhcnIgPSBbMSwgMiwgM11cblxuY29uc3QgaW5jID0gKG46IG51bWJlcikgPT4gbiArIDFcbmNvbnN0IGluY0FsbCA9IG1hcFZhbHVlcyhpbmMpXG5jb25zdCByZXN1bHQgPSBpbmNBbGwoYXJyKVxuc2hvdygnaW5jQWxsKGFyciknLCByZXN1bHQpIiwibGFuZ3VhZ2UiOiJ0cyIsInZlcnNpb24iOjF9)
